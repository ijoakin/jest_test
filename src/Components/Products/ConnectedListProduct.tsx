import { connect } from "react-redux";
import { Products } from ".";
import { ProductsState } from "../Store/Reducers/ProductReducer";
import { ListProducts } from "./ListProducts";
import { Product } from "./product";

export interface ListProductsProps {
  products: Product[];
  total: number;
}

export const mapStateToProps = (state: ProductsState): ListProductsProps => {
  return {
    products: state.listProducts,
    total: state.total
  };
};

export default connect(mapStateToProps)(ListProducts);