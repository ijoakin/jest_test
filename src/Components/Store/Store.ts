import { createStore } from 'redux';
import { ProductReducer } from './Reducers/ProductReducer';

export const store = createStore(ProductReducer);