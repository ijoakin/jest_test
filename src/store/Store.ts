import { createStore } from 'redux';
import { initialState, ProductReducer } from './Reducers/ProductReducer';
import {composeWithDevTools} from 'redux-devtools-extension';

const composeEnhancers = composeWithDevTools({});

export const store = createStore(ProductReducer, initialState, composeEnhancers());