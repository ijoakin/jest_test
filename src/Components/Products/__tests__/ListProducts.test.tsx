import { Provider } from "react-redux";
import rendered from "react-test-renderer";
import {
  ProductReducer,
  ProductsState,
} from "../../../store/Reducers/ProductReducer";
import { store } from "../../../store/Store";
import ConnectedListProduct, { mapStateToProps } from "../ConnectedListProduct";
import delay from "redux-saga";

const getMockState = (): ProductsState => {
  return {
    product: { description: "", id: 0, price: 0 },
    listProducts: [
      {
        description: "description",
        id: 10,
        price: 25,
      },
      {
        description: "description2",
        id: 11,
        price: 25,
      },
      {
        description: "description3",
        id: 12,
        price: 25,
      },
      {
        description: "description4",
        id: 13,
        price: 25,
      },
    ],
    total: 0,
  };
};

jest.mock("../../../Services/ProductService.ts");

describe("List Products Component Tests", () => {
  let state: ProductsState;

  beforeEach(() => {
    state = getMockState();
  });

  it("Should render as expected", () => {
    const tree = rendered
      .create(
        <Provider store={store}>
          <ConnectedListProduct></ConnectedListProduct>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Should total be equal to zero when the component is rendered", async () => {
    const tree = rendered.create(
      <Provider store={store}>
        <ConnectedListProduct></ConnectedListProduct>
      </Provider>
    );

    await delay();

    const instance = tree.root;
    const component = instance.findByProps({ id: "txtTotal" });

    const text = component.props.value; // this going to get the text

    expect(text).toEqual(0);
  });
  it("Should calculate using the component", async () => {
    const tree = rendered.create(
      <Provider store={store}>
        <ConnectedListProduct></ConnectedListProduct>
      </Provider>
    );

    await delay();

    const instance = tree.root;
    const component = instance.findByProps({ id: "txtTotal" });

    const text = component.props.value; // this going to get the text

    expect(text).toEqual(0);
  });
  it("Should calculate getting values from mock", async () => {
    const tree = rendered.create(
      <Provider store={store}>
        <ConnectedListProduct></ConnectedListProduct>
      </Provider>
    );

    const instance = tree.root;
    const component = instance.findByProps({ id: "txtTotal" });

    const text = component.props.value; // this going to get the text

    expect(text).toEqual(0);
  });

  it("Should Props be correct", () => {
    const componentProps = mapStateToProps(state);

    expect(componentProps.total).toEqual(0);
    expect(componentProps.products.length).toEqual(4);
  });
  it("Should total calculation", () => {
    const newstate = ProductReducer(state, {
      type: "CALCULATE_TOTAL",
      payload: state,
    });

    expect(newstate.total).toEqual(100);
  });
});
