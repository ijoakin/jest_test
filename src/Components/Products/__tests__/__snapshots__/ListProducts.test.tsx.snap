// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Products Component Tests Should render as expected 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table"
  >
    <thead>
      <th>
        Id
      </th>
      <th>
        Description
      </th>
      <th>
        Price
      </th>
    </thead>
    <tbody />
  </table>
  <div>
    <button
      id="btnGetProduct"
      onClick={[Function]}
    >
      Get Product
    </button>
  </div>
  <div>
    <button
      id="btnGetTotal"
      onClick={[Function]}
    >
      Get Total
    </button>
    <input
      className="divtxtTotal"
      id="txtTotal"
      type="text"
      value={0}
    />
  </div>
</div>
`;
